sources/packages/api/src/services/disbursement-schedule-service/disbursement-schedule-service.ts

```diff
@@ -0,0 +1,120 @@
+import { Injectable } from "@nestjs/common";
+import { CustomNamedError } from "../../utilities";
+import { Connection, Repository } from "typeorm";
+import {
+  APPLICATION_NOT_FOUND,
+  APPLICATION_NOT_VALID,
+  INVALID_OPERATION_IN_THE_CURRENT_STATUS,
+  SequenceControlService,
+} from "..";
+import { RecordDataModelService } from "../../database/data.model.service";
+import {
+  Application,
+  ApplicationStatus,
+  DisbursementSchedule,
+  DisbursementValue,
+} from "../../database/entities";
+import { Disbursement } from "./disbursement-schedule.models";
+
+const DISBURSEMENT_DOCUMENT_NUMBER_SEQUENCE_GROUP =
+  "DISBURSEMENT_DOCUMENT_NUMBER";
+
+/**
+ * Service layer for Student Application disbursement schedules.
+ */
+@Injectable()
+export class DisbursementScheduleService extends RecordDataModelService<DisbursementSchedule> {
+  private readonly applicationRepo: Repository<Application>;
+  constructor(
+    connection: Connection,
+    private readonly sequenceService: SequenceControlService,
+  ) {
+    super(connection.getRepository(DisbursementSchedule));
+    this.applicationRepo = connection.getRepository(Application);
+  }
+
+  /**
+   * Create the disbursements while the application is 'In Progress'.
+   * Ensures that the application do not have any disbursements records
+   * and creates the disbursements and values altogether.
+   * ! This methods should be called for the first time when the
+   * ! disbursements are being added to the Student Application.
+   * ! Once the Student Application already has disbursements, another
+   * ! scenarios must be considered, for instance, if some amount of money
+   * ! was already released to the student.
+   * @param applicationId application id to associate the disbursements.
+   * @param disbursements array of disbursements and values to be created.
+   * @returns created disbursements.
+   */
+  async createDisbursementSchedules(
+    applicationId: number,
+    disbursements: Disbursement[],
+  ): Promise<DisbursementSchedule[]> {
+    const application = await this.applicationRepo
+      .createQueryBuilder("application")
+      .select([
+        "application.id",
+        "application.applicationStatus",
+        "disbursementSchedules.id",
+      ])
+      .leftJoin("application.disbursementSchedules", "disbursementSchedules")
+      .where("application.id = :applicationId", { applicationId })
+      .getOne();
+
+    if (!application) {
```

Comment:
Is it required to have throw custom error status for different application status and disbursements schedule? Or can we have the checks in the query to select only the application in the right status?

Reply:
That is a super valid question. I followed what we did in another area and I would say that this is not what we need to follow everywhere.
The reason that I decided to be more specific for this particular method is:
1. This method will be called most likely only by the workflow and it is really unlikely to be called when the application is in the wrong state, which means that the record retrieved would be almost the time right;
2. In case of something really unusual happens and this method throws that particular exception, it could be used by technical people to troubleshoot this, likely a TBTB person. It means that, if this method would be likely returning the exception message to a Student or Institution user, I would say that only a more generic error message would be better and I would include the status in the query.
Does it make sense?